### ISSUE TITLE
Add more probe control parameters

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#12866]
Resurrected from #66 and #761.

Probe currently supports the following control parameters:
```
	// Length of time before health checking is activated.  In seconds.
	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty" description:"number of seconds after the container has started before liveness probes are initiated; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes"`
	// Length of time before health checking times out.  In seconds.
	TimeoutSeconds int64 `json:"timeoutSeconds,omitempty" description:"number of seconds after which liveness probes timeout; defaults to 1 second; see http://releases.k8s.io/HEAD/docs/user-guide/pod-states.md#container-probes"`
```

We should add the following common control parameters (matching, for instance, [AppEngine's parameters](https://cloud.google.com/appengine/docs/managed-vms/config#health_checking) and [GCE's parameters](https://cloud.google.com/compute/docs/reference/latest/httpHealthChecks)):
```
// How often to probe
PeriodSeconds *int64
// Minimum consecutive successes for the probe to be considered successful after having reached the FailureThreshold; defaults to 1; must be 1 for liveness 
SuccessThreshold *int
// Minimum consecutive failures for the probe to be considered failed; defaults to 1
FailureThreshold *int
```

At some point we'll probably want to add an enable/disable boolean, since it's convenient to be able to disable liveness probes without losing the parameter settings, but that's less critical than the above.

**Comment by dchen1107** [comment]
cc/ @timstclair @davidopp

**Comment by thockin** [comment]
Small point - let's choose "interval" or "period" and use it everywhere.  I
have found (and perhaps contributed to inconsistency).

**Comment by yujuhong** [comment]
To do this, we need to decouple probing from pod syncing.
We can either wake up the pod worker for probing only, or we can add a separate component solely for probing. 
The decoupling also needed by my pod lifecycle event proposal (as described the [TODO](https://github.com/yujuhong/kubernetes/blob/kubelet_proposal/docs/proposals/pod-lifecycle-event-generator.md#todos) list) since periodic sync will be eliminated eventually.

**Comment by bgrant0607** [comment]
Good catch, @thockin. Period it is. Updated issue description.

**Comment by karlkfi** [comment]
In the vein of probe improvements: I'm not sure InitialDelaySeconds is really desirable as currently documented/implemented. I suspect a better option would be one starts probing immediately upon creation but delays interpretation of failed probes as failure until after the designated delay. This would allow probed resources to be found to be alive/ready ASAP, while waiting up to the configured time to be marked as dead/failed.

Additionally, I think having the prober itself do retries is a mistake that degrades the resolution of the probe to 3 x TimeoutSeconds. It seems likely that some of the pod state transition logic leaked into the prober to avoid pronouncing a pod failed after 1 failed attempt. But I'm not sure that SuccessThreshold or FailureThreshold belong in the prober at all. Those are configurations for state transition logic, not a data point gatherer. The prober should be agnostic to the state transition logic.

I have an example impl of the re-imagined InitialDelaySeconds definition in the component prober from #12324: https://github.com/mesosphere/kubernetes/blob/component-registration/pkg/component/prober.go

I suspect we should go a step further tho and make state transitioning its own layer. 

While I do think PeriodSeconds should be added to the prober, I suspect the values selected for the SuccessThreshold, FailureThreshold, and TimeoutSeconds all will be impacted by the value selected for the PeriodSeconds. A shorter timeout (and removal/reduction of retries) allows you to reduce the period, and a shorter period might require you to increase the thresholds. So maybe you don't want to have a single threshold config shared by multiple independently configurable timeouts/periods.

**Comment by timstclair** [comment]
@karlkfi Are you commenting on the API design, implementation, or both? I think that FailureThreshold could make the need for retries obsolete, but this is debatable since a retry would be immediate while the failure threshold would wait for the next probePeriod before retrying.

WRT initialDelaySeconds, what is the motivation for wanting to know whether a Pod is alive before the delay?

**Comment by karlkfi** [comment]
@timstclair little of column A, little of column B. The name InitialDelaySeconds is ambiguous that you have to go by the documented/commented description, which describes the implementation. Technically we could change the implementation without changing the API, except that the API would then do something it didn't do before. As long as no one relies on a pod not recieving a probe too soon, it should be fine. But we'd have to also change the description of the field, which is tantamount to changing the API.

Alternatively, we could support both options, but that makes the API more complicated and then you have a naming problem...

If we add SuccessThreshold or FailureThreshold to probe we have to remove retries, because they could be set to different values. We'd have to track the probe results outside of the prober, which is fine by me. However, I think tracking it outside of the prober means that there's a missing abstraction layer, which should probably be represented in the API. That layer would handle phase/state transition.

Trying to re-use the Probe API model objects for Components brought me to the realization that probing should be independent from phase, because probing only provides discrete data points, each of which is fallible. Probing alone cannot determine the phase of a pod or a component. The phase is determined by applying some logical analysis, an algorithm, to the probe data. Then a determination about the phase is made. The more configurable that algorithm is, the more complex it is. The more complex it is, the more reason there is to separate it from the probing data collection. The more separate it is in implementation, the more reason there is that the API should be separate too.

As for your question about InitialDelaySeconds, think about Readiness rather than Liveness. Lets say your container can take anywhere between 5s and 60s to come up, depending on network bandwidth, latency, cache state, dependency readiness, etc. Do you want to wait 60s to know that your container is ready or do you want to know ASAP? As-is you'd have to put a 60s InitialDelaySeconds. Any sooner and your container would get killed for not being ready and restarted, potentially thrashing, not giving enough time for it for come up. And then even if your app is ready in 5s, you wont know until 55s later, if you're relying on k8s to tell you.

**Comment by timstclair** [comment]
@karlkfi I agree with what you're saying with regards to the state transition from an implementation standpoint, but I'd worry that representing the state transition logic separately from the probe could add unecessary complexity to the API. I'm too green to make any strong statements about the API structure though.

Also, I believe containers are not restarted based on Readiness state, only Liveness. So in your example it could just sit in the Readiness Failed (which I believe is the [default](http://kubernetes.io/v1.0/docs/user-guide/pod-states.html#container-probes) while waiting for the initial delay anyway). I was wondering if there was a use case for knowing liveness before the initial delay that couldn't be replaced with a readiness check.

**Comment by karlkfi** [comment]
@timstclair If Probe were only used by pod containers, then the state transition config for pod containers could be in Probe, for simplicity and least change, but adding it to Probe means that the Probe model can only ever be used by other resources that share the same phase transition logic and support the same config options. @bgrant0607 was trying to convince me to not add Phase with my Components PR, but without Phase, SuccessThreshold and FailureThreshold don't make sense.

I haven't proposed a good alternative API that's reverse compatible and solves all my concerns, so it's hard to say "yeah, that sounds good". But that's partly because I haven't hashed out the use cases yet.

I'd also point to SuccessThreshold needing to be set to 1 for liveness as indicating that Probe is the wrong layer of abstraction for that option.

The fact that we already support liveness as distinct from readiness sets k8s apart from every other health checker impl I've ever seen. It makes sense, but it means we can't just blindly copy existing interface options.

**Comment by bgrant0607** [comment]
I don't recommend using initialDelaySeconds with readiness probes. We could not share the parameter definitions across readiness and liveness probes, at the cost of a little more code duplication.

I don't see what SuccessThreshold and FailureThreshold have to do with Phase. They are about translation of probe results to condition status.

We could potentially change killing behavior wrt liveness condition status. Pass failure threshold => kill. Don't kill again unless transition to success and then failure again?

**Comment by bgrant0607** [comment]
WRT liveness vs. readiness separation: Internally these checks are implemented by 2 entirely different systems. The disadvantage to implementing readiness somewhere else is that it becomes harder for other components to take advantage of the information, which means they have to implement their own checks, which need to be specified in more places (load balancers, application updaters, node updaters, monitoring systems, ...), placing more load on the monitored application, etc.

**Comment by karlkfi** [comment]
Perhaps I misinterpreted what you intend to do with the SuccessThreshold and FailureThreshold. I'm not sure what "2 entirely different systems" you mean, either, unless you mean docker vs rkt. I was under the impression that the [kublet prober](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go) was used for both liveness and readiness probing of pod containers. I assumed you were going to use SuccessThreshold/FailureThreshold to replace the hardcoded 3 retries loop. 

The prober usage is harder to follow than its implementation, but I was under the impression the prober liveness and readiness results were being used to determine the phase AND condition of the pod container, rather than just updating the condition and using the conditions over time to update the phase.

I see [dockertools.Manager using the prober](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/manager.go#L1604), but that code is particularly hard to follow, especially being in an 1800+ line file. I suspect that adding a layer of logical abstraction there that is common for both docker and rocket is exactly what it needs...

**Comment by timstclair** [comment]
FWIW I'm working on pulling the prober out of the manager as a prerequisite to fixing this issue.

**Comment by karlkfi** [comment]
@timstclair glad to hear it. send me a link when you're done and i'll review it

**Comment by timstclair** [comment]
Questions about defaults:

- `PeriodSeconds` - To maintain full backwards compatibility, the default value should be the kubelet sync period, but that is [variable](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L215). Should we stick with this behavior (fall back to sync period), or can we set the default to be the [default sync period](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L206) (10s)

- `FailureThreshold` - The [current prober implementation](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L40) will retry on failure up to 3 times. If we get rid of this behavior, instead using `FailureThreshold` with a default value of 3, it will be a slight change in behavior since it will wait `PeriodSeconds` between each retry (the current implementation retries immediately). Is this change in behavior acceptable? If not, should we introduce another `MaxRetries` parameter?

**Comment by karlkfi** [comment]
I tend to figure the reverse compatibility contract applies to the interface, not necessarily the implementation. If there's some depending on the old impl, they were probably doing something wrong.

**Comment by yujuhong** [comment]
> PeriodSeconds - To maintain full backwards compatibility, the default value should be the kubelet sync period, but that is variable. Should we stick with this behavior (fall back to sync period), or can we set the default to be the default sync period (10s)

I think apiserver should supply the default, and that would not be based on kubelet's configuration at all. We never promised (i.e., explicitly stated) the probing period, so backward compatibility should not be an issue.

**Comment by timstclair** [comment]
Great, that's a lot cleaner :) What about the prober retries? Should we keep them? Another option would be to retry on prober error, but not failure.

**Comment by yujuhong** [comment]
> The prober usage is harder to follow than its implementation, but I was under the impression the prober liveness and readiness results were being used to determine the phase AND condition of the pod container, rather than just updating the condition and using the conditions over time to update the phase.

I am a bit lost on the "phase" vs "condition" argument. Readiness is used to set the "Ready" bit in the container status, as well as affect the Pod ready condition. It would not affect the pod phase.
Liveness, on the other hand, would only cause containers to be restarted on failures.

**Comment by yujuhong** [comment]
> FailureThreshold - The current prober implementation will retry on failure up to 3 times. If we get rid of this behavior, instead using FailureThreshold with a default value of 3, it will be a slight change in behavior since it will wait PeriodSeconds between each retry (the current implementation retries immediately). Is this change in behavior acceptable? If not, should we introduce another MaxRetries parameter?

@bgrant0607, do we intend to change the existing behavior?

**Comment by bgrant0607** [comment]
The intent is to match GCE and GAE health-checking APIs in functionality, though not necessarily the same default values, etc. Those APIs are fairly mature at this point, and mirror internal health-checking APIs, as well.

I would prefer not to add a maxRetries parameter to the API. 

Neither the interval nor the retry policy can be configured currently -- they are implementation-dependent. The interval is also load-dependent. We are changing that behavior, at least. We're not going to make the default interval 1 second.

In order for this change to break an app, that app would already have to be failing probes. Users might actually want to be made aware of that.

So, I think we should change the behavior.

It would be useful to get the opinions of a couple people who might be aware of valid assumptions this could break, though. cc @roberthbailey @liggitt

**Comment by bgrant0607** [comment]
FWIW, phase is not relevant to this at all.

**Comment by shaklee3** [comment]
Any chance sub-second probes will ever be a feature? One second can be a very long time when a failure happens.

**Comment by thockin** [comment]
Can you open a new bug for this?

**Comment by shaklee3** [comment]
Sure. Thanks
